### PR TITLE
Reduce gcc memory usage on deal.II development tester

### PR DESCRIPTION
**Pull request by gassmoeller**
This PR reduces the required memory for the tester that uses the deal.II development version. With the previous official deal.II image we often saw the github action run out of memory. We now build our own deal.II image once a day and use the same settings as for the testers that use deal.II releases.

### CODE REVIEW

File: .github/workflows/linux.yml
Comment (tjhei) [inline .github/workflows/linux.yml:135]: \

**Comment by tjhei** [comment]
I would try -j2 first because ninja uses n+1=3 by default.

**Comment by gassmoeller** [comment]
Interesting, I didnt know that. `man ninja` says the default is #CPUS, but it does not specify how it computes #CPUS. I can try 2 if 1 works.

**Comment by gassmoeller** [comment]
Looks like 2 processes is too much and leads to the crash. Let's see if the tester runs if I use a custom build docker image (the same way we use a custom image for the releases). I imagine that might use less memory, because it only includes the deal.II dependencies that ASPECT needs. We could build the custom image automatically, just as we build the main ASPECT docker image. (maybe once daily or every few days).

**Comment by gassmoeller** [comment]
With the custom docker image the tester succeeds and compiling is also more than twice as fast as the official deal.II docker image (likely because less dependencies are included). I think this is something we should pursue, but I want to set up a github action first that automatically rebuilds the deal.II image in a reasonable time. I will create a follow-up PR some time next week. Do we need a quick fix for the dev tester?

**Comment by tjhei** [comment]
> Interesting, I didnt know that. `man ninja` says the default is #CPUS

If you do ``ninja -h`` it will tell you the default. My machine has 8 cores, 16 threads, and ninja uses 18.

**Comment by tjhei** [comment]
> (likely because less dependencies are included).

How can fewer deal.II dependencies result in faster ASPECT compilation with less memory?
Is the compiler the same?

**Comment by gassmoeller** [comment]
> How can fewer deal.II dependencies result in faster ASPECT compilation with less memory?
> Is the compiler the same?

Well, I do not know the details, but I would assume more included headers, less removed lines by the preprocessor, and a larger deal.II library means more work for the compiler and linker. This is consistent with the difference in size of the precompiled header file between geodynamics/aspect-tester (635MB) vs dealii/dealii:master-focal (851MB). One additional influence may be that for geodynamics/aspect-tester we remove debug symbols, and compile deal.II debug with -O3 instead of -O0. 

Anyway to compare the results here are two outputs of /usr/bin/time on the same machine with exactly the same cmake options and use of ninja -j2 in the two different docker containers:

geodynamics/aspect-tester:latest:
```
[117/117] Linking CXX executable aspect
773.03user 43.34system 6:58.67elapsed 194%CPU (0avgtext+0avgdata 3766996maxresident)k
7968inputs+1994408outputs (26major+16075361minor)pagefaults 0swaps
```

dealii/dealii:master-focal:
```
[117/117] Linking CXX executable aspect
1458.11user 79.70system 13:04.87elapsed 195%CPU (0avgtext+0avgdata 5970244maxresident)k
8inputs+11733408outputs (158067major+27066299minor)pagefaults 0swaps
```

So not only does it take twice the time for dealii/dealii it also require nearly 6GB RAM instead of 3.76GB.

**Comment by tjhei** [comment]
That difference is impressive. How do you want to go about this now? How do we update deal.II?

**Comment by bangerth** [comment]
How do we want to proceed here?

**Comment by tjhei** [comment]
Discussed this with @gassmoeller: I will prepare an alternative running on my jenkins instead.

**Comment by gassmoeller** [comment]
Actually I tested this a bit more on my fork. Building a new dealii dev docker container for this tester just takes about 2 hours (https://github.com/gassmoeller/aspect/runs/7657669991?check_suite_focus=true) so we could easily do it as a github action nightly. The benefit would be that we have the three github actions build on containers that have consistent build parameters for deal.II (all with minimal dependencies and -O3 for debug mode).

I agree that using the official deal.II image for your Jenkins tester would add value (because then we also test the official image), but we could also do that in addition to the changes in this PR. In the current state this PR should work and fix the build problems for the deal.II-dev github action.

**Comment by tjhei** [comment]
If you want to set it up this way, please go ahead.

Did you see that the most recent dealii master is working again (but we have failing tests)?

**Comment by gassmoeller** [comment]
> If you want to set it up this way, please go ahead.

Ok, I interpret that as a ready to merge?

> Did you see that the most recent dealii master is working again (but we have failing tests)?

Interesting, it seems gmg mesh deformation tests are failing with floating point exceptions. But this is currently using a deal.II dev from a few weeks ago when I set up this PR first. Let me merge this and tomorrow we will have a fresh deal.II image.

File: .github/workflows/docker_tester.yml
Comment (tjhei) [inline .github/workflows/docker_tester.yml:1]: I think the naming of this job and filename can be improved.

**Comment by tjhei** [comment]
No, the Deal-master is working on main without this PR merged. See any other new PR. We might not need to merge this PR after all.

We are working on the FPEs in #4905.

**Comment by gassmoeller** [comment]
Oh interesting. Well I am ok with closing this then, we can always come back to it if it becomes a problem in the future again.

**Comment by gassmoeller** [comment]
Let me reopen this. @tjhei this should exactly fix #4961 as we discussed this morning. I renamed the file, do you think the new name is good enough? I am open for suggestions.

File: .github/workflows/build_tester_base_image.yml
Comment (tjhei) [inline .github/workflows/build_tester_base_image.yml:40]: Should we add deal.II or "master" or something to the name? It shows up in the description of the runs on each pull request.

File: contrib/docker/tester/Dockerfile
Comment (tjhei) [inline contrib/docker/tester/Dockerfile:31]: Does it make sense to keep this folder intact or maybe introduce subfolders for the different versions?

**Review by tjhei — APPROVED** [review]

**Comment by tjhei** [comment]
Can you also please update the PR description?

File: contrib/docker/tester/Dockerfile
Comment (gassmoeller) [reply contrib/docker/tester/Dockerfile:31]: This is the name candi chooses inside the install directory. I do not know if I can change that.

**Review by gassmoeller — COMMENTED** [review]
Updated the PR description and container name. I do not know if I can do anything about the deal.II install directory inside the container though.

**Comment by gassmoeller** [comment]
If you agree with the container name I can build it once manually to make the tester succeed (it currently fails because the container will only be built once this PR is merged).

File: contrib/docker/tester/Dockerfile
Comment (tjhei) [reply contrib/docker/tester/Dockerfile:31]: I am sorry, I meant "does it make sense to keep ``contrib/docker/tester/Dockerfile`` as is or introduce a 9.4 and a master version"?

File: contrib/docker/tester/build.sh
Comment (tjhei) [inline contrib/docker/tester/build.sh:5]: why is this ``latest`` and not ``focal-dealii-master``?

File: contrib/docker/tester/Dockerfile
Comment (gassmoeller) [reply contrib/docker/tester/Dockerfile:31]: Ah I see. Well they are pretty identical and easy to change. I would prefer not creating another copy that can get out of date.

File: contrib/docker/tester/build.sh
Comment (gassmoeller) [reply contrib/docker/tester/build.sh]: My bad I did not check that file. Fixed now.

**Comment by tjhei** [comment]
Do you want to manually push ``geodynamics/aspect-tester:focal-dealii-master`` so that the master tester succeeds? Otherwise, feel free to merge.

**Review by tjhei — APPROVED** [review]

**Comment by tjhei** [comment]
Looks like #4973 did the trick!

**Pull request merged**
